allennlp/data/dataset_readers/dataset_utils/span_utils.py

```diff
@@ -0,0 +1,125 @@
+from typing import List, Tuple, Callable
+
+from allennlp.data.dataset_readers.dataset_utils.ontonotes import TypedStringSpan
+from allennlp.data.tokenizers.token import Token
+
+def enumerate_spans(sentence: List[Token],
+                    offset: int = 0,
+                    max_span_width: int = None,
+                    min_span_width: int = None,
+                    filter_function: Callable[[List[Token]], bool] = None) -> List[Tuple[int, int]]:
+    """
+    Given a sentence, return all token spans within the sentence. Spans are `inclusive`.
+    Additionally, you can provide a maximum and minimum span width, which will be used
+    to exclude spans outside of this range.
+
+    Finally, you can provide a function mapping ``List[Token] -> bool``, which will
+    be applied to every span to decide whether that span should be included. This
+    allows filtering by length, regex matches, pos tags or any Spacy ``Token`` attribute.
+
+    Parameters
+    ----------
+    sentence : ``List[Token]``, required.
+        The sentence to generate spans for.
+    offset : ``int``, optional (default = 0)
+        A numeric offset to add to all span start and end indices. This is helpful
+        if the sentence is part of a larger structure, such as a document, which
+        the indices need to respect.
+    max_span_width : ``int``, optional (default = None)
+        The maximum length of spans which should be included.
+    min_span_width : ``int``, optional (default = None)
+        The minimum length of spans which should be included.
+    """
+    max_span_width = max_span_width or len(sentence)
+    min_span_width = min_span_width or 0
+    filter_function = filter_function or bool
+    spans: List[Tuple[int, int]] = []
+
+    for start_index in range(len(sentence)):
+        last_end_index = min(start_index + max_span_width, len(sentence))
+        first_end_index = min(max(start_index, start_index + min_span_width - 1), len(sentence))
```

Comment:
I think `min_span_width` should default to 1, not 0.  I don't think it makes much sense to have a bunch of empty spans at various places.  And if the default is 1, you can remove this inner max.

Reply:
Ah, ok, yeah, I was confused by a few pieces of this (like why the `bool` filter wasn't doing what I thought).  This max effectively changes the `min_span_width` to 1 if it was set at 0.  So just set it to 1 by default.